README.md
# PynAbi

Python package to easily create [Abinit](https://www.abinit.org/) input files.

```cmd
pip install pynabi
```

## Example

```python
from pynabi import createAbi, DataSet, AbIn, AbOut
from pynabi.kspace import CriticalPointsOf, BZ, SymmetricGrid, UsualKShifts, Path
from pynabi.calculation import ToleranceOn, EnergyCutoff, MaxSteps, SCFMixing, NonSelfConsistentCalc
from pynabi.crystal import Atom, FluoriteLike
from pynabi.occupation import OccupationPerBand
from pynabi.units import eV, nm

# create manually an atom -> Atom(<Z>, <pseudo potential name>)
# or using sensible defaults as follows
Zr = Atom.of("Zr")  # Z=40 and pseudos located at "Zr.psp8"
Oxy = Atom.of("O")  # Z=8 and pseudos located at "O.psp8"

# base dataset with common variables
base = DataSet(
    AbOut("./scf/scf"),                             # prefix for output files
    AbIn().PseudoPotentials("./pseudos/PBE-SR"),    # folder with pseudo potentials

    FluoriteLike(Zr, Oxy, 0.5135*nm),               # creates AtomBasis and Lattice of a crystal like fluorite
                                                    # with lattice constant 0.5135nm

    SymmetricGrid(BZ.Irreducible, UsualKShifts.FCC) # easily define kptopt, ngkpt, nshiftk, kpt
        .ofMonkhorstPack(4),

    SCFMixing(density=True).Pulay(10),              # scf cycle with Pulay mixing of the density 
                                                    # based on the last 10 iteration

    ToleranceOn.EnergyDifference(1e-6),             # expressively define the tolerance
    MaxSteps(30)                                    # nstep
)

# set the default energy unit in eV (from now on)
eV.setAsReference()

# datasets to see the convergenge as a function of energy
sets = [DataSet(EnergyCutoff(8.0 + i*0.25)) for i in range(0,17)]

# final non-self-consistent round to find bands 
bands = DataSet(
    NonSelfConsistentCalc(),
    ToleranceOn.WavefunctionSquaredResidual(1e-12),
    AbIn().ElectronDensity(sets[-1]),                   # get the electron density from the last dataset
    OccupationPerBand(2.0, repeat=8),                   # same number of bands (max 8) for each k point
    Path.auto(10, "GXWKGLUWLK", CriticalPointsOf.FCC)   # easily define a path in the k-space   
)

with open("./out.txt", 'w') as f:
    f.write(createAbi(base, *sets, bands))
```

<details>
<summary><b>Output</b></summary>

```txt
ndtset 18

# Atoms definition
ntypat 2
znucl 40 8
pseudos "Zr.psp8, O.psp8"

# Common DataSet
natom 3
typat 1 2 2
xred 0 0 0   0.3333333333333333 0.3333333333333333 0.3333333333333333   0.6666666666666666 0.6666666666666666 0.6666666666666666
outdata_prefix "./scf/scf"
pp_dirpath "./pseudos/PBE-SR"
scalecart 0.5135 0.5135 0.5135 nm
rprim 0.5 0.5 0.0   0.0 0.5 0.5   0.5 0.0 0.5
kptopt 1
nshiftk 4
shiftk 0.5 0.5 0.5 0.5 0.0 0.0 0.0 0.5 0.0 0.0 0.0 0.5
ngkpt 4 4 4
iscf 17
npulayit 10
toldfe 1e-06
nstep 30

# DataSet 1
ecut1 8.0 eV

# DataSet 2
ecut2 8.25 eV

# DataSet 3
ecut3 8.5 eV

# DataSet 4
ecut4 8.75 eV

# DataSet 5
ecut5 9.0 eV

# DataSet 6
ecut6 9.25 eV

# DataSet 7
ecut7 9.5 eV

# DataSet 8
ecut8 9.75 eV

# DataSet 9
ecut9 10.0 eV

# DataSet 10
ecut10 10.25 eV

# DataSet 11
ecut11 10.5 eV

# DataSet 12
ecut12 10.75 eV

# DataSet 13
ecut13 11.0 eV

# DataSet 14
ecut14 11.25 eV

# DataSet 15
ecut15 11.5 eV

# DataSet 16
ecut16 11.75 eV

# DataSet 17
ecut17 12.0 eV

# DataSet 18
iscf18 -2
tolwfr18 1e-12
getden18 17
occopt18 0
occ18 8*2.0
band18 8
kptopt18 -9
kptbounds18 0 0 0  0.0 0.5 0.5  0.25 0.75 0.5  0.375 0.75 0.375  0 0 0  0.5 0.5 0.5  0.25 0.625 0.625  0.25 0.75 0.5  0.5 0.5 0.5  0.375 0.75 0.375
ndivsm18 10
```

</details>

## Features

 - Multi dataset support
 - Helper functions for common crystal structures (caesium chloride, rock-salt, fluorite, zincblende, wurtzite, nickeline, HCP)
 - Registered critical points of and methodsto create lattices of CUB, BCC, FCC, HEX, TET, BCT, ORC, ORCC
 - Smooth experience in defining the k-points
 - Handy management of [file handling variables](https://docs.abinit.org/variables/files/)
 - Almost full covarage of [basic input variables](https://docs.abinit.org/variables/basic/) (missing nbandhf, symrel, tnons, wvl_hgrid)
 - Partial coverage of [ground state variables](https://docs.abinit.org/variables/gstate/)
 - Partial coverage of [relaxation variables](https://docs.abinit.org/variables/rlx/)
 - _More to come..._

## Why PynAbi over pure Abinit files?

If you're a very experienced Abinit user who knows its variables and their possible values (and associated meaning), then this package probably isn't for you.
For all other users, on top of all the aforementioned features, here's a list of reasons why you could find PynAbi useful:

 1. You have all the power of coding to generate Abinit instructions, e.g. reusability, loops to generate datasets programmatically
 2. It provides some useful presets and helper functions/methods that allows you to skip to the fun part of the simulation
 3. Under the hood, it checks for the validity of variable values *before* starting Abinit
 4. It makes use of expressive declarations and definition, leading to readable and comprehensible istructions
 5. If you're using a code editor (with autocompletition), you'll get suggetions of all the possible options in a more natural language 

Some prior knowledge of Abinit (and DFT in general) is nonetheless needed to fully understand what actually is to be simulated: this package only provides a handier way to generate the required files.

## Documentation

Although currently severely incomplete, a wiki is available on the [GitHub repository Wiki](https://github.com/Fedesky25/pynabi/wiki). Ideally a wiki for this package shouldn't be needed if your editor has some autocompletition: the aim is to document every class and function with a detailed description taken from the relevant parts of ABINIT documentation. 

CHANGELOG.md
# PynAbi Changelog

## 0.1.2

 * Feature: `SCFMixing` and `SCFDirectMinimization` require `Tolerance` to be specified
 * Feature: when `NonSelfConsistentCal` is not explicitly set, PynAbi will check the existence of `Tolerance` (Abinit implicitly assumes a SCF calculation)
 * Documentation: for `Smearing` and `Metal`
 * Fix: `typat`, `natom`, `nband` spelled correctly
 * Fix: compatibility check also on base dataset
 * Fix: `SymmetricGrid` now correctly writes `kptopt` and `shiftk` 

## 0.1.1

 * Fix: `CanDelay` in same dataset of one of its `Delayed` properties must delay that property
 * Feature: you can now append stampable to groups of datsets to cut down repeated code

## 0.1.0

 * __Breaking__: refactor of `kspace` -> there are now three specialized (mutually exclusive) classes `ManualGrid`, `SymmetricGrid`, `AutomaticGrid`, `Path`
 * Refactor using `IndexedWithDefault` to cut down boilerplate code
 * Refactor of the delaying code: `DelayedInfo` has been added and a delayed value now can be of type different from input
 * `SCFDirectMinimization` does not take steps as argument (already taken by `MaxSteps`)
 * Reject delayed value only if already defined in same dataset

## 0.0.3 and previous

 * File handling thorugh `AbIn` and `AbOut`
 * Atom basis and lattices with many helper functions
 * Definition of the reciprocal space
 * Occupation of bands and spin types
 * Basic relaxation: implemention of `ionmov`, `occopt` with their asscoated variables
 * Units of energy and length of Abinit
 * (non) SCF calculation
 * Partial check for imcompatible definitions and values

pynabi/kspace.py
from ._common import Vec3D as _V, Stampable as _Stmp, _pos_int
from enum import Enum as _E
from typing import Dict as _dict, Any as _any, Union as _union, Tuple as _tuple, Iterable as _iter


__all__ = ["BZ", "CriticalPointsOf", "manual", "SymmetricGrid", "UsualKShifts", "path"]


class BZ(_E):
    Irreducible = 1
    Half = 2
    Full = 3
    NoTimeReversal = 4


class CriticalPointsOf(_E):
    """
    Critical points of (some) Brillouin zones\n
    Taken from http://lampx.tugraz.at/~hadley/ss1/bzones/
    """
    CUB = {
        'R': _V(0.5, 0.5, 0.5),
        'X': _V(0.0, 0.5, 0.0),
        'M': _V(0.5, 0.5, 0.0)
    }
    BCC = {
        'H': _V(-0.5, 0.5, 0.5),
        'P': _V.uniform(0.25),
        'N': _V(0.0, 0.5, 0.0)
    }
    FCC = {
        'X': _V(0.0, 0.5, 0.5),
        'L': _V.uniform(0.5),
        'W': _V(0.25, 0.75, 0.5),
        'U': _V(0.25, 0.625, 0.625),
        'K': _V(0.375, 0.75, 0.375)
    }
    HEX = {
        'A': _V(0,0,1/2),
        'K': _V(2/3,1/3,0),
        'H': _V(2/3,1/3,1/2),
        'M': _V(1/2,0,0),
        'L': _V(1/2,0,1/2)
    }
    TET = {
        'X': (0.5,0.0,0.0),
        'M': (0.5,0.5,0.0),
        'Z': (0.0,0.0,0.5),
        'R': (0.5,0.0,0.5),
        'A': _V.uniform(0.5)
    }
    BCT = {
        'X': _V(0.5,0.0,0.0),
        'Z': _V(0.5,0.5,-0.5),
        'N': _V(0.0,0.5,0.0),
        'P': _V.uniform(0.25)
    }
    ORC = {
        'X': _V(0.5,0.0,0.0),
        'Y': _V(0.0,0.5,0.0),
        'Z': _V(0.0,0.0,0.5),
        'T': _V(0.0,0.5,0.5),
        'U': _V(0.5,0.0,0.5),
        'S': _V(0.5,0.5,0.5),
        'R': _V.uniform(0.5)
    }
    ORCC = {
        'Y': _V(0.5,0.5,0.0),
        'Y': _V(-0.5,0.5,0.0),
        'Z': _V(0.0,0.0,0.5),
        'T': _V.uniform(0.5),
        'T': _V(-0.5,0.5,0.5),
        'S': _V(0.0,0.5,0.0),
        'R': _V(0.0,0.5,0.5),
    }


class UsualKShifts(_E):
    Unshifted = (_V.zero(),)
    Default = (_V.uniform(0.5),)
    BCC = (_V.uniform(0.25), _V.uniform(-0.25))
    FCC = (_V.uniform(0.5), _V(0.5,0.0,0.0), _V(0.0,0.5,0.0), _V(0.0,0.0,0.5))
    HEX = (_V(1.0,0.0,0.0), _V(-0.5,0.8660254037844386,0.0), _V(0.0,0.0,1.0))


class KSpaceDefinition(_Stmp):
    def __init__(self, option: int, props: _dict[str, _any]) -> None:
        super().__init__()
        self.kptop = option
        self.props = props
    
    def stamp(self, index: int):
        s = index or ''
        head = '' if self.kptop == 100 else f"kptopt{s} {self.kptop}\n"
        body = '\n'.join(f"{k}{s} {v}" for k,v in self.props.items())
        return f"{head}{body}"


def manual(*points: _V, normalize: float = 1.0):
    assert normalize >= 1, "k-points normalization faction cannot be lower than 1"
    return KSpaceDefinition(0, {
        "nkpt": len(points),
        "kpt": "   ".join(str(v) for v in points),
        "kptnrm": normalize
    })


class SymmetricGrid:
    def __init__(self, symmetry: BZ, shifts: _union[_tuple[_V,...], UsualKShifts] = ()):
        self._sy = symmetry
        if type(shifts) is UsualKShifts:
            self._sh = shifts.value
        elif type(shifts) is tuple:
            assert all(type(v) is _V for v in shifts), "K Shifts must be vectors"
            self._sh = shifts
        else:
            raise TypeError("Invalid type of k shifts")
    
    def _u(self, n: str, v: str):
        assert len(self._sh) > 0, "There must be at least one k shift"
        return KSpaceDefinition(self._sy.value, {
            "nshiftk": len(self._sh),
            "shiftk": "   ".join(str(s) for s in self._sh),
            n: v
        })
    
    def ofMonkhorstPack(self, a: int, b: _union[int,None] = None, c: _union[int,None] = None):
        b = a if b is None else b
        c = a if c is None else c
        assert _pos_int(a) and _pos_int(b) and _pos_int(c), "Number of k points in Monkhorst-Pack grid must be a positive integer"
        return self._u("ngkpt", f"{a} {b} {c}")
    
    def fromSuperLattice(self, a: _V, b: _V, c: _V):
        return self._u("kptrlatt", f"{a} {b} {c}")
    

    def automatic(self, length: float = 30.0):
        assert length > 0, "Real space length used for automatic k grid must be positive"
        return KSpaceDefinition(self._sy.value, { "kptrlen": length })


def setMPGridPointsNumber(number: _union[int, _tuple[int, int, int]]):
    if type(number) is tuple:
        assert len(number) == 3, "numbers of k points must be three"
        assert all(_pos_int(v) for v in number), "numbers of k points must be positive integers"
        return KSpaceDefinition(100, { "ngkpt": f"{number[0]} {number[1]} {number[2]}" })
    else:
        assert _pos_int(number), "Number of k points (per primitive) must be a positive integer"
        return KSpaceDefinition(100, { "ngkpt": f"{number} {number} {number}" })


def path(divisions: _union[int,_tuple[int,...]], points: _union[str,_iter[_union[str,_V]]], pointSet: _union[CriticalPointsOf,_dict[str,_V]] = {}):
    s: dict[str,_V] = pointSet.value if type(pointSet) is CriticalPointsOf else pointSet  # type: ignore
    b: list[_V] = []
    for p in points:
        if type(p) is str:
            for c in p:
                if c == 'G':
                    b.append(_V.zero())
                else:
                    assert c in s, f"(critical) point '{c}' is not defined"
                    b.append(s[c])
        elif type(p) is _V:
            b.append(p)
        else:
            raise TypeError(f"Invalid type of k-path point (got {type(p)})")
    assert len(b) > 1, "Number of boundaries must be at least 2 (i.e. one segment)"
    
    if type(divisions) is tuple:
        assert len(divisions) == len(b)-1, f"lenght of division must be equal to number of segments: got {len(divisions)} instead of {len(b)-1}"
        assert all(_pos_int(v) for v in divisions), "Number of division per segment must be all positive"
        return KSpaceDefinition(1-len(b), { 
            "kptbounds": "   ".join(str(v) for v in b), 
            "ndivk": ' '.join(str(d) for d in divisions)
        })
    else:
        assert _pos_int(divisions), "number of division (for smallest segment) must be positive"
        return KSpaceDefinition(1-len(b), { 
            "kptbounds": "   ".join(str(v) for v in b), 
            "ndivsm": divisions
        })